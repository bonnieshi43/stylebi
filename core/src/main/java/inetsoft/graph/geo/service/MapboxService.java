/*
 * inetsoft-core - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package inetsoft.graph.geo.service;

import inetsoft.sree.SreeEnv;

import javax.json.*;
import javax.net.ssl.HttpsURLConnection;
import java.awt.*;
import java.io.IOException;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.List;

public class MapboxService extends WebMapService {
   public MapboxService() {
      user = SreeEnv.getProperty("mapbox.user");
      token = SreeEnv.getProperty("mapbox.token");
      style = SreeEnv.getProperty("mapbox.style");
   }

   public MapboxService(boolean orgScoped) {
      user = SreeEnv.getProperty("mapbox.user", false, orgScoped);
      token = SreeEnv.getProperty("mapbox.token", false, orgScoped);
      style = SreeEnv.getProperty("mapbox.style", false, orgScoped);
   }

   public MapboxService(String user, String token, String style) {
      this.user = user;
      this.token = token;
      this.style = style;
   }

   /**
    * Load a static image using bbox.
    */
   @Override
   public Image loadImage(double lonMin, double lonMax, double latMin, double latMax,
                          int imgw, int imgh)
         throws IOException
   {
      // mapbox static image api:
      // https://api.mapbox.com
      // /styles/v1/{username}/{style_id}/static/{bbox}/{width}x{height}?access_token=...
      String urlstr = String.format("https://api.mapbox.com" +
                                    "/styles/v1/%s/%s/static/[%f,%f,%f,%f]/%dx%d?access_token=%s",
                                 user, style, lonMin, latMin, lonMax, latMax, imgw, imgh, token);
      return WebMapService.loadFromURL(urlstr);
   }

   /**
    * Get a list of all map styles for the user.
    */
   public List<MapboxStyle> getStyles(boolean withDefault) throws IOException {
      List<MapboxStyle> styles = new ArrayList<>();
      getStyles("https://api.mapbox.com/styles/v1/" + user + "?access_token=" + token, styles);

      if(withDefault) {
         styles.add(0, MapboxStyle.DEFAULT);
      }

      return styles;
   }

   private void getStyles(String urlstr, List<MapboxStyle> styles) throws IOException {
      URL url = new URL(urlstr);
      URLConnection conn = url.openConnection();

      try {
         String next = conn.getHeaderField("next");
         JsonReader parser = Json.createReader(conn.getInputStream());
         JsonArray arr = (JsonArray) parser.read();

         arr.stream().map(v -> new MapboxStyle(v)).forEach(v -> styles.add(v));

         if(next != null) {
            getStyles(next, styles);
         }
      }
      catch(IOException ex) {
         if(conn instanceof HttpsURLConnection) {
            if(((HttpsURLConnection) conn).getResponseCode() == 403) {
               throw new RuntimeException("Access token requires permission to STYLES:LIST secret scope");
            }
         }

         throw ex;
      }
   }

   /**
    * Get the tile size. One tile is required to show world at lowest zoom level.
    */
   @Override
   public int getTileSize() {
      return 512;
   }

   @Override
   public int getMaxSize() {
      return 1280;
   }

   @Override
   public boolean isDiscreteZoom() {
      return false;
   }

   private String user;
   private String token;
   private String style;
}
