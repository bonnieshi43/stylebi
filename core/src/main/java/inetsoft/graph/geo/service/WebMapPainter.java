/*
 * inetsoft-core - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package inetsoft.graph.geo.service;

import inetsoft.graph.*;
import inetsoft.graph.coord.Coordinate;
import inetsoft.graph.coord.GeoCoord;
import inetsoft.graph.geo.GeoProjection;

import java.awt.*;
import java.awt.geom.Rectangle2D;
import java.awt.image.BufferedImage;
import java.io.IOException;

public class WebMapPainter implements BackgroundPainter {
   public WebMapPainter() {
   }

   public WebMapPainter(WebMapService service) {
      this.service = service;
   }

   /**
    * Get the Web map service used by this painter.
    */
   public WebMapService getWebMapService() {
      return service;
   }

   @Override
   public void prepareCoordinate(GeoCoord coord, int width, int height) {
      service.prepareCoordinate(coord, width, height);
   }

   @Override
   public void paint(Graphics2D g, VGraph graph) throws IOException {
      Rectangle2D plot = graph.getPlotBounds();
      Coordinate coord = graph.getCoordinate();
      Image bg = getImage(graph);

      if(bg != null && coord instanceof GeoCoord) {
         double padding = getPadding((GeoCoord) coord);

         if(padding != 0) {
            BufferedImage img = (BufferedImage) bg;
            // fetch the sub image without the padding (which is used for panning).
            int subx = (int) (plot.getWidth() * padding);
            int suby = (int) (plot.getHeight() * padding);
            bg = img.getSubimage(subx, suby, (int) plot.getWidth(), (int) plot.getHeight());
         }
      }

      if(bg != null) {
         GraphTool.drawImage(g, bg, plot.getX(), plot.getY(), plot.getWidth(), plot.getHeight());
      }
   }

   /**
    * Get the image used for drawing background, including padding.
    */
   @Override
   public Image getImage(VGraph graph) throws IOException {
      Image bg = graph.getPlotBackground();

      if(bg == null && graph.getCoordinate() instanceof GeoCoord) {
         bg = loadImage(graph);
         graph.setPlotBackground(bg);
      }

      return bg;
   }

   private Image loadImage(VGraph graph) throws IOException {
      GeoCoord coord = (GeoCoord) graph.getCoordinate();
      GeoProjection proj = coord.getProjection();
      Rectangle2D bbox = coord.getBBox();
      Rectangle2D plot = coord.getVGraph().getPlotBounds();
      double plotw = plot.getWidth();
      double ploth = plot.getHeight();

      if(Double.isNaN(bbox.getWidth()) || Double.isNaN(bbox.getHeight())) {
         return null;
      }

      // adding padding to the image for panning.
      // mapbox uses bbox to calculate the zoom and center. if width/height is larger than
      // bbox, the image is expended on each size to fill in the space.
      // !!! lon/lat in web map is unprojected and x/y in coordinate (e.g. scale) are projected.

      double padding = getPadding(coord);
      int imgw = (int) (plotw * (1 + padding * 2));
      int imgh = (int) (ploth * (1 + padding * 2));
      // bbox corresponds to image (web)
      double xmin = bbox.getMinX() - bbox.getWidth() * padding;
      double xmax = bbox.getMaxX() + bbox.getWidth() * padding;
      double ymin = bbox.getMinY() - bbox.getHeight() * padding;
      double ymax = bbox.getMaxY() + bbox.getHeight() * padding;
      double yrange = ymax - ymin;
      double xrange = xmax - xmin;

      // empty space around the edge in case the lat/lon is out of range
      int topBlank = 0, bottomBlank = 0, leftBlank = 0, rightBlank = 0;
      // max for mapbox is 85.0511
      final double MAXY = proj.projectY(proj.getYMax());

      while(xmin < -360) {
         xmin += 360;
         xmax += 360;
      }

      while(xmax > 360) {
         xmin -= 360;
         xmax -= 360;
      }

      if(ymax > MAXY) {
         double diff = ymax - MAXY;
         topBlank = (int) (diff * imgh / yrange);
         ymax = MAXY;
      }

      if(ymin < -MAXY) {
         double diff = -MAXY - ymin;
         bottomBlank = (int) (diff * imgh / yrange);
         ymin = -MAXY;
      }

      if(xmin < -360) {
         double diff = -360 - xmin;
         leftBlank = (int) (diff * imgw / xrange);
         xmin = -360;

         // keep center at the same place if arbitrary bounds not supported.
         if(service.isDiscreteZoom()) {
            rightBlank = leftBlank;
         }
      }

      // if map doesn't support arbitrary bounds, we must maintain the same center for the
      // image. otherwise the map would not line up with the plot.
      if((topBlank > 0 || bottomBlank > 0) && service.isDiscreteZoom()) {
         topBlank = bottomBlank = Math.max(topBlank, bottomBlank);
      }

      double webw = imgw - leftBlank - rightBlank;
      double webh = imgh - topBlank - bottomBlank;
      int maxSize = service.getMaxSize();

      // if size > MAX_SIZE, reduce it on both sides to fit
      if(webw > maxSize) {
         double range = xmax - xmin;
         double adj = (webw - maxSize) / 2; // reduce on each side
         xmin += adj * range / webw;
         xmax -= adj * range / webw;
         webw = maxSize;
         leftBlank += adj;
      }

      if(webh > maxSize) {
         double range = ymax - ymin;
         double adj = (webh - maxSize) / 2; // reduce on each side
         ymin += adj * range / webh;
         ymax -= adj * range / webh;
         webh = maxSize;
         topBlank += adj;
      }

      double lonMin = proj.inverseX(xmin);
      double lonMax = proj.inverseX(xmax);
      double latMin = proj.inverseY(ymin);
      double latMax = proj.inverseY(ymax);

      Image img = service.loadImage(lonMin, lonMax, latMin, latMax, (int) webw, (int) webh);

      // fill in blank space at top and bottom
      if(topBlank > 0 || leftBlank > 0 || bottomBlank > 0 || rightBlank > 0) {
         BufferedImage img2 = new BufferedImage(imgw, imgh, BufferedImage.TYPE_INT_ARGB);
         Graphics2D g = (Graphics2D) img2.getGraphics();
         g.setColor(Color.WHITE);
         g.fillRect(0, 0, imgw, imgh);
         g.drawImage(img, leftBlank, topBlank, null);
         g.dispose();
         img = img2;
      }

      return img;
   }

   private WebMapService service;
}
