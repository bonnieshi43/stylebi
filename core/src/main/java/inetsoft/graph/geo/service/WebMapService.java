/*
 * inetsoft-core - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package inetsoft.graph.geo.service;

import inetsoft.graph.coord.GeoCoord;
import inetsoft.sree.SreeEnv;
import inetsoft.uql.viewsheet.graph.MapInfo;
import inetsoft.util.MessageException;

import javax.imageio.ImageIO;
import javax.net.ssl.HttpsURLConnection;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.io.Serializable;
import java.net.URL;
import java.net.URLConnection;

/**
 * This is the interface for web map services.
 */
public abstract class WebMapService implements Serializable {
   static BufferedImage loadFromURL(String urlstr) throws IOException {
      URLConnection conn = null;

      try {
         URL url = new URL(urlstr);
         conn = url.openConnection();
         return ImageIO.read(conn.getInputStream());
      }
      catch(IOException ex) {
         if(conn instanceof HttpsURLConnection) {
            int code = ((HttpsURLConnection) conn).getResponseCode();

            if(code == 402) {
               throw new WebMapLimitException();
            }
            else if(code == 403) {
               throw new MessageException("Web map access key is invalid: " + ex.getMessage());
            }
         }

         throw ex;
      }
   }

   /**
    * Get the current web map service.
    * @param mapboxStyle mapbox style or null to use the default.
    */
   public static WebMapService getWebMapService(String mapboxStyle) {
      String service = SreeEnv.getProperty("webmap.service");

      if(MapInfo.MAPBOX.equals(service)) {
         String mapboxUser = SreeEnv.getProperty("mapbox.user");
         String mapboxToken = SreeEnv.getProperty("mapbox.token");

         if(mapboxStyle == null) {
            mapboxStyle = SreeEnv.getProperty("mapbox.style");
         }

         return new MapboxService(mapboxUser, mapboxToken, mapboxStyle);
      }
      else if(MapInfo.GOOGLE.equals(service)) {
         return new GoogleMapsService();
      }

      return null;
   }

    /**
    * Load a static image using bbox.
    */
   public abstract Image loadImage(double lonMin, double lonMax, double latMin, double latMax,
                   int imgw, int imgh)
      throws IOException;

   /**
    * Get the tile size. One tile is required to show world at lowest zoom level.
    */
   public abstract int getTileSize();

   /**
    * Get the maximum size for the image.
    */
   public abstract int getMaxSize();

   /**
    * Check if this web map service only support discrete zoom levels or arbitrary zoom.
    */
   public abstract boolean isDiscreteZoom();

   /**
    * Get the maximum zoom level for discrete zoom.
    */
   public int getMaxZoomLevel() {
      return 22;
   }

   /**
    * Get the zoom level for the map. This may be explicitly set through setZoomLevel()
    * or calculated by prepareCoordinate().
    */
   public int getZoomLevel() {
      return zoomLevel;
   }

   /**
    * Set the zoom level for the map. This will be used for the map if this map service
    * requires discrete zoom level.
    */
   public void setZoomLevel(int zoomLevel) {
      this.zoomLevel = zoomLevel;
   }

   /**
    * Check and adjust latitude/longitude range if necessary.
    */
   public void prepareCoordinate(GeoCoord coord, int width, int height) {
      // no op
   }

   private int zoomLevel = -1;
}
