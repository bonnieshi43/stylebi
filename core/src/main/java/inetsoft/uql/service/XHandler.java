/*
 * This file is part of StyleBI.
 * Copyright (C) 2024  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package inetsoft.uql.service;

import inetsoft.mv.MVSession;
import inetsoft.report.XSessionManager;
import inetsoft.report.composition.execution.AssetDataCache;
import inetsoft.uql.*;
import inetsoft.uql.schema.XVariable;
import inetsoft.uql.util.*;
import inetsoft.util.DataCacheVisitor;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.security.Principal;
import java.util.Enumeration;
import java.util.Hashtable;

/**
 * XHandler defines the API for query handlers. A handler is responsible
 * for a particular protocol. All interfaces to a data source, including
 * connection, meta data retrieval, and query execution are processed
 * through a handler.
 *
 * @version 5.1, 9/20/2003
 * @author InetSoft Technology Corp
 */
public abstract class XHandler implements java.io.Serializable {
   /**
    * Execute the query.
    * @param params parameters for query.
    * @return the result as a hierarchical tree.
    */
   public abstract XNode execute(XQuery query, VariableTable params,
                                 Principal user, DataCacheVisitor visitor)
      throws Exception;

   /**
    * Connect to the data source.
    * @param datasource the data source name.
    * @param params parameters for connection.
    */
   public abstract void connect(XDataSource datasource, VariableTable params)
      throws Exception;

   /**
    * Test the data source.
    * @param datasource the data source name.
    * @param params parameters for connection.
    * @throws Exception
    */
   public abstract void testDataSource(XDataSource datasource,
      VariableTable params) throws Exception;

   /**
    * Build the meta data of this data source as a XNode tree. This
    * method will rebuild the meta data tree everytime it's called.
    * The meta data should be cached by the caller.
    * @param mtype meta data type, defined in each data source.
    * @return return the root node of the meta data tree.
    */
   public abstract XNode getMetaData(XNode mtype) throws Exception;

   /**
    * Close the data source connection.
    */
   public abstract void close() throws Exception;

   /**
    * Set the session associated with this handler.
    */
   public void setSession(Object session) {
      this.session = session;
   }

   /**
    * Set the repository associated with this handler.
    */
   public void setRepository(XRepository repository) {
      this.repository = repository;
   }

   /**
    * Clear the internal cached states.
    */
   public void reset() {
      qparamMap.clear();
   }

   /**
    * Clear the internal cached states of a query.
    *
    * @param query the specified query
    */
   public void reset(XQuery query) {
      qparamMap.remove(getQueryKey(query));
   }

   /**
    * Apply max rows to the resultset.
    */
   public XNode postProcess(XNode root, XQuery query, VariableTable params)
      throws Exception {
      int max = query.getMaxRows();

      if(params != null && params.get(XQuery.HINT_MAX_ROWS) != null) {
         max = Integer.parseInt(params.get(XQuery.HINT_MAX_ROWS).toString());
      }

      if(max == 0) {
         return root;
      }

      if(root instanceof XSequenceNode) {
         int amax = 0;

         while(root.getChildCount() > max) {
            root.removeChild(root.getChildCount() - 1);
            amax = max;
         }

         if(amax > 0) {
            ((XSequenceNode) root).setAppliedMaxRows(amax);
         }
      }
      else if(root instanceof XTableNode) {
         root = new XTableNodeWrapper((XTableNode) root);
         ((XTableNodeWrapper) root).setMaxRows(max);
      }

      return root;
   }

   /**
    * Get the variable table to serve as base table. This contains all
    * query variables, whose values are calculated at runtime.
    */
   protected void prepareVariableTable(XQuery xquery, VariableTable vars) {
      String qkey = getQueryKey(xquery);
      VariableTable vt = (VariableTable) qparamMap.get(qkey);

      if(vt == null) {
         qparamMap.put(qkey, vt = new CachedVariableTable());
         // find all un-set variables in the query variables
         Enumeration keys = xquery.getVariableNames();

         while(keys.hasMoreElements()) {
            String name = (String) keys.nextElement();
            XVariable var = xquery.getVariable(name);
            vt.put(name, var);
         }
      }

      // @by larryl, remove old chain to avoid stack overflow
      vars.removeBaseTable(CachedVariableTable.class);
      vars.addBaseTable(vt);
      vars.setSession(session);
   }

   protected boolean isQueryCached(XQuery query, VariableTable vars, Principal user,
                                   DataCacheVisitor visitor) throws Exception
   {
      if(visitor == null) {
         return false;
      }

      if(AssetDataCache.isDebugData()) {
         return false;
      }

      final String key = getQueryKey(query, vars, user);
      return visitor.visitCache(key);
   }

   public void removeQueryCache(XQuery query, VariableTable vars, Principal user, Class<?> type) throws Exception {
      final String key = getQueryKey(query, vars, user) + type;
      XSessionManager.getSessionManager().removeCacheData(key);
   }

   /**
    * Get a query key for query caching.
    */
   public String getQueryKey(XQuery query, VariableTable qvars, Principal user) throws Exception {
      // @by jasons, bug1249326025439, if the _USER_ or _ROLES_ variables are
      // applied by a VPM, the values for these variables must be added to the
      // variable table so that the query keys are distinct for different users
      // or roles. If not, the VPM conditions may not be applied and the
      // incorrect data may be returned
      VariableTable kvars = new VariableTable();

      // copy only the parameters used in query to kvars so changes in parameters unrelated
      // to the query won't cause the cache to be missed. (60409)
      for(Enumeration e = query.getVariableNames(); e.hasMoreElements();) {
         String vname = (String) e.nextElement();

         if("_USER_".equals(vname) && !kvars.contains(vname)) {
            kvars.put(vname, user.getName());
         }
         else if("_ROLES_".equals(vname) && !kvars.contains(vname)) {
            kvars.put(vname, XUtil.getUserRoles(user));
         }
         else if("_GROUPS_".equals(vname) && !kvars.contains(vname)) {
            kvars.put(vname, XUtil.getUserGroups(user));
         }
         else {
            kvars.put(vname, qvars.get(vname));
         }
      }

      // we always use the cache since even if caching is turned off, we
      // still cache the data within one report (one XSession.execute)
      // ds + "__" + "query xml" + "__" + "qvars" + "__" + "queryMaxRow"
      StringWriter buffer = new StringWriter();
      PrintWriter writer = new PrintWriter(buffer);
      XDataSource source = query.getDataSource();
      writer.print(source == null ? "null" : source.getFullName());
      writer.print("__");
      query.writeXML(writer);
      writer.print("__");
      kvars.printKey(writer);

      if(user != null && Identity.UNKNOWN_USER.equals(user.getName())) {
         writer.print("__");
         writer.print(user.toString());
      }

      if(query.dependsOnMVSession()) {
         final MVSession mvSession = MVSession.getCurrentSession();
         writer.print(mvSession == null ? "" : mvSession.getSessionDiscriminator());
      }

      writer.flush();
      return buffer.toString();
   }

   /**
    * Get the key for qparamMap.
    */
   private static String getQueryKey(XQuery query) {
      String name = query.getName();
      XDataSource dx = query.getDataSource();
      String folder = query.getFolder();

      return name + "_" + folder + "_" + ((dx == null) ? "" : dx.getFullName());
   }

   // marker class for clean up
   private static class CachedVariableTable extends VariableTable {
   }

   protected Object session = null;
   protected XRepository repository = null;
   private Hashtable qparamMap = new Hashtable(); // XQuery -> VariableTable
}
