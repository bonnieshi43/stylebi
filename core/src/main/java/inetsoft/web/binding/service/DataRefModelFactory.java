/*
 * This file is part of StyleBI.
 * Copyright (C) 2024  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package inetsoft.web.binding.service;

import inetsoft.uql.erm.DataRef;
import inetsoft.web.binding.drm.DataRefModel;

public abstract class DataRefModelFactory<R extends DataRef, M extends DataRefModel> {
   /**
    * Gets the data ref class supported by this factory.
    *
    * @return the data ref class.
    */
   public abstract Class<R> getDataRefClass();

   /**
    * Creates a new model instance for the specified dataRef.
    *
    * @param dataRef the dataRef interface.
    * @return a new model.
    */
   public abstract M createDataRefModel(R dataRef);
}
