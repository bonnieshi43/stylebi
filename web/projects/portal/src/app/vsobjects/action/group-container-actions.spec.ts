/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright Â© 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { TestUtils } from "../../common/test/test-utils";
import { ComposerContextProviderFactory, ViewerContextProviderFactory } from "../context-provider.service";
import { GroupContainerActions } from "./group-container-actions";

describe("GroupContainerActions", () => {
   const createModel = () => TestUtils.createMockVSGroupContainerModel("GroupContainer1");

   it("check status of menu actions of group container in composer", () => {
      const expectedMenu = [
         [
            { id: "group properties", visible: true },
            { id: "group show-format-pane", visible: true }
         ],
         [
            { id: "vs-object copy", visible: true },
            { id: "vs-object cut", visible: true },
            { id: "vs-object remove", visible: true },
            { id: "vs-object group", visible: true },
            { id: "vs-object ungroup", visible: true }
         ],
         [
            { id: "vs-object bring-forward", visible: true },
            { id: "vs-object bring-to-front", visible: true },
            { id: "vs-object send-backward", visible: true },
            { id: "vs-object send-to-back", visible: true }
         ]
      ];

      const model = createModel();
      const actions = new GroupContainerActions(model, ComposerContextProviderFactory());
      const menuActions = actions.menuActions;
   });

   it("check status of menu actions of group container in viewer and preview", () => {
      const expectedMenu = [
         [
            { id: "group properties", visible: false },
            { id: "group show-format-pane", visible: false }
         ],
         [
            { id: "vs-object copy", visible: false },
            { id: "vs-object cut", visible: false },
            { id: "vs-object remove", visible: false },
            { id: "vs-object group", visible: false },
            { id: "vs-object ungroup", visible: false }
         ],
         [
            { id: "vs-object bring-forward", visible: false },
            { id: "vs-object bring-to-front", visible: false },
            { id: "vs-object send-backward", visible: false },
            { id: "vs-object send-to-back", visible: false }
         ]
      ];

      //check status in viewer
      const model = createModel();
      const actions1 = new GroupContainerActions(model, ViewerContextProviderFactory(false));
      const menuActions1 = actions1.menuActions;

      expect(menuActions1).toMatchSnapshot();

      //check status in preview
      const model2 = createModel();
      const actions2 = new GroupContainerActions(model2, ViewerContextProviderFactory(true));
      const menuActions2 = actions2.menuActions;

      expect(menuActions2).toMatchSnapshot();
   });

   //Bug #19986 should not display menu action when as data tip component
   it("should not display menu action when as data tip component", () => {
      const dataTipService: any = { isDataTip: jest.fn() };
      dataTipService.isDataTip.mockImplementation(() => true);
      const model = createModel();
      const actions = new GroupContainerActions(model, ViewerContextProviderFactory(false), false, null, dataTipService);
      const menuActions = actions.menuActions;

      expect(menuActions.length).toBe(0);
   });
});