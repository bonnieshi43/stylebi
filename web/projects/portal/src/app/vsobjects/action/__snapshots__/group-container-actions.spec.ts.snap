// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupContainerActions check status of menu actions of group container in viewer and preview 1`] = `
Array [
  AssemblyActionGroup {
    "actions": Array [
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
    ],
    "icon": [Function],
    "label": [Function],
  },
  AssemblyActionGroup {
    "actions": Array [
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
    ],
    "icon": [Function],
    "label": [Function],
  },
  AssemblyActionGroup {
    "actions": Array [
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
    ],
    "icon": [Function],
    "label": [Function],
  },
]
`;

exports[`GroupContainerActions check status of menu actions of group container in viewer and preview 2`] = `
Array [
  AssemblyActionGroup {
    "actions": Array [
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
    ],
    "icon": [Function],
    "label": [Function],
  },
  AssemblyActionGroup {
    "actions": Array [
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
    ],
    "icon": [Function],
    "label": [Function],
  },
  AssemblyActionGroup {
    "actions": Array [
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
      Object {
        "action": [Function],
        "enabled": [Function],
        "icon": [Function],
        "id": [Function],
        "label": [Function],
        "visible": [Function],
      },
    ],
    "icon": [Function],
    "label": [Function],
  },
]
`;
